routing/fake_ending.cpp

```diff
@@ -0,0 +1,79 @@
+#include "routing/fake_ending.hpp"
+
+#include "routing/edge_estimator.hpp"
+#include "routing/index_graph.hpp"
+#include "routing/world_graph.hpp"
+
+#include "geometry/distance.hpp"
+
+#include "base/math.hpp"
+
+#include <functional>
+
+using namespace routing;
+using namespace std;
+
+namespace
+{
+using EstimatorFn = function<double(m2::PointD const &, m2::PointD const &)>;
+
+Junction CalcProjectionToSegment(Junction const & begin, Junction const & end,
+                                 m2::PointD const & point, EstimatorFn const & estimate)
+{
+  m2::ProjectionToSection<m2::PointD> projection;
+
+  projection.SetBounds(begin.GetPoint(), end.GetPoint());
+  auto const projectedPoint = projection(point);
+  auto const distBeginToEnd = estimate(begin.GetPoint(), end.GetPoint());
+
+  double constexpr kEpsMeters = 2.0;
+  if (my::AlmostEqualAbs(distBeginToEnd, 0.0, kEpsMeters))
+    return Junction(projectedPoint, begin.GetAltitude());
+
+  auto const distBeginToProjection = estimate(begin.GetPoint(), projectedPoint);
+  auto const altitude = begin.GetAltitude() + (end.GetAltitude() - begin.GetAltitude()) *
+                                                  distBeginToProjection / distBeginToEnd;
+  return Junction(projectedPoint, altitude);
+}
+
+FakeEnding MakeFakeEndingImpl(Junction const & backJunction, Junction const & frontJunction,
```

Comment:
Please, consider to get rid of std::function here.

Reply:
done